flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/RegionPartitionReleaseStrategyTest.java

```diff
@@ -0,0 +1,149 @@
+
+		regionPartitionReleaseStrategy.vertexFinished(onlyIntermediateVertexId);
+		final List<IntermediateResultPartitionID> partitionsToRelease = regionPartitionReleaseStrategy.vertexFinished(onlySinkVertexId);
+		assertThat(partitionsToRelease, contains(onlySourceResultPartitionId));
+	}
+
+	@Test
+	public void notReleasePartitionsIfDownstreamRegionIsNotFinished() {
+		final List<TestingSchedulingExecutionVertex> producers = testingSchedulingTopology.addExecutionVertices().finish();
+		final List<TestingSchedulingExecutionVertex> consumers = testingSchedulingTopology.addExecutionVertices().withParallelism(2).finish();
+		testingSchedulingTopology.connectAllToAll(producers, consumers).finish();
+
+		final ExecutionVertexID onlyProducerVertexId = producers.get(0).getId();
+		final ExecutionVertexID consumerVertex1 = consumers.get(0).getId();
+		final ExecutionVertexID consumerVertex2 = consumers.get(1).getId();
+
+		final Set<PipelinedRegion> pipelinedRegions = pipelinedRegionsSet(
+			PipelinedRegion.from(onlyProducerVertexId),
+			PipelinedRegion.from(consumerVertex1, consumerVertex2));
+
+		final RegionPartitionReleaseStrategy regionPartitionReleaseStrategy = new RegionPartitionReleaseStrategy(testingSchedulingTopology, pipelinedRegions);
+
+		final List<IntermediateResultPartitionID> partitionsToRelease = regionPartitionReleaseStrategy.vertexFinished(consumerVertex1);
+		assertThat(partitionsToRelease, is(empty()));
+	}
+
+	@Test
+	public void toggleVertexFinishedUnfinished() {
+		final List<TestingSchedulingExecutionVertex> producers = testingSchedulingTopology.addExecutionVertices().finish();
+		final List<TestingSchedulingExecutionVertex> consumers = testingSchedulingTopology.addExecutionVertices().withParallelism(2).finish();
+		testingSchedulingTopology.connectAllToAll(producers, consumers).finish();
+
+		final ExecutionVertexID onlyProducerVertexId = producers.get(0).getId();
+		final ExecutionVertexID consumerVertex1 = consumers.get(0).getId();
+		final ExecutionVertexID consumerVertex2 = consumers.get(1).getId();
+
+		final Set<PipelinedRegion> pipelinedRegions = pipelinedRegionsSet(
+			PipelinedRegion.from(onlyProducerVertexId),
+			PipelinedRegion.from(consumerVertex1, consumerVertex2));
+
+		final RegionPartitionReleaseStrategy regionPartitionReleaseStrategy = new RegionPartitionReleaseStrategy(testingSchedulingTopology, pipelinedRegions);
+
+		regionPartitionReleaseStrategy.vertexFinished(consumerVertex1);
+		regionPartitionReleaseStrategy.vertexFinished(consumerVertex2);
+		regionPartitionReleaseStrategy.vertexUnfinished(consumerVertex2);
+
+		final List<IntermediateResultPartitionID> partitionsToRelease = regionPartitionReleaseStrategy.vertexFinished(consumerVertex1);
+		assertThat(partitionsToRelease, is(empty()));
+	}
+
+	private static Set<PipelinedRegion> pipelinedRegionsSet(final PipelinedRegion... pipelinedRegions) {
```

Comment:
might be further deduplication via 
`pipelinedRegionsSet(ExecutionVertexID producerIds, ExecutionVertexID... consumerIds)`

Reply:
I prefer the current abstraction. 


In this signature it is unclear where the pipelined region boundaries are:

```
pipelinedRegionsSet(ExecutionVertexID producerIds, ExecutionVertexID... consumerIds)
```

```
pipelinedRegionsSet(onlyProducerVertexId, consumerVertex1, consumerVertex2);
```